packages/shaders/src/shader-mount.ts

```diff
@@ -143,28 +189,83 @@ export class ShaderMount {
   private renderScale = 1;
-  /** Resize handler for when the container div changes size and we want to resize our canvas to match */
+
+  /** Resize handler for when the parent element changes size and we want to resize our canvas to match */
   private handleResize = () => {
-    const clientWidth = this.parentElement.clientWidth;
-    const clientHeight = this.parentElement.clientHeight;
+    const parentClientWidth = this.parentElement.clientWidth;
+    const parentClientHeight = this.parentElement.clientHeight;
     const maxResolution = this.maxResolution;
+
     // Note we render at 2X even for 1x screens because it gives a much smoother looking result
     const pixelRatio = Math.max(2, window.devicePixelRatio);
-    // Render scale prevents the virtual resolution from going beyond our maxium and then multiplies by pixelRatio (so at least 2X rendering)
-    this.renderScale = Math.min(1, maxResolution / Math.max(clientWidth, clientHeight)) * pixelRatio;
 
-    let newWidth = clientWidth * this.renderScale;
-    let newHeight = clientHeight * this.renderScale;
-    if (this.canvasElement.width !== newWidth || this.canvasElement.height !== newHeight) {
-      this.canvasElement.width = newWidth;
-      this.canvasElement.height = newHeight;
+    // Pattern shaders come with an infinite world size by default – trim it to the parent element
+    let canvasClientWidth = this.worldWidth === Infinity ? parentClientWidth : this.worldWidth;
+    let canvasClientHeight = this.worldHeight === Infinity ? parentClientHeight : this.worldHeight;
+    const worldAspectRatio = canvasClientWidth / canvasClientHeight;
+
+    if (this.worldFit !== 'crop') {
+      const referenceWidth = Math.max(canvasClientWidth, parentClientWidth);
+      const referenceHeight = Math.max(canvasClientHeight, parentClientHeight);
+      const clamp = this.worldFit === 'cover' ? Math.max : Math.min;
+      canvasClientWidth = worldAspectRatio * clamp(referenceWidth / worldAspectRatio, referenceHeight);
+      canvasClientHeight = canvasClientWidth / worldAspectRatio;
+    }
+
+    // Canvas offsets relative to the corresponding sides of the parent element
+    const widthOffset = parentClientWidth - canvasClientWidth;
+    const heightOffset = parentClientHeight - canvasClientHeight;
+    const canvasLeft = widthOffset * this.worldOriginX;
+    const canvasTop = heightOffset * this.worldOriginY;
+    const canvasRight = widthOffset * (1 - this.worldOriginX);
+    const canvasBottom = heightOffset * (1 - this.worldOriginY);
+
+    const baseViewportWidth = parentClientWidth - Math.max(0, canvasLeft) - Math.max(0, canvasRight);
+    const baseViewportHeight = parentClientHeight - Math.max(0, canvasTop) - Math.max(0, canvasBottom);
+
+    // Scale the render according to the device pixel ratio, but not beyond the maximum resolution
+    this.renderScale = Math.min(pixelRatio, maxResolution ** 2 / (baseViewportWidth * baseViewportHeight * pixelRatio));
```

Comment:
@StephenHaney it didn't quite make sense to me that a 2560x10px stripe would be downsized to save some cycles, but 1920x1920 would be fine

I updated the max resolution to be total physical pixels (including the pixel ratio now). Does this make sense?

Reply:
Can we close this PR and can we port this idea to the other one?

Reply:
Yeah! Actually thought about this today and realized we still have a few big flaws in how we calculate `renderScale`. Will get back to that tomorrow and open a separate PR for setting shader dpi